analytics/src/test/java/com/segment/analytics/EdgeFunctionMiddlewareTest.kt

```diff
@@ -0,0 +1,141 @@
+
+    @Mock
+    lateinit var integrationBar: Integration<Void>
+
+    @Before
+    fun setUp() {
+        MockitoAnnotations.initMocks(this)
+        Analytics.INSTANCES.clear()
+        TestUtils.grantPermission(RuntimeEnvironment.application, Manifest.permission.INTERNET)
+        val projectSettings =
+                ValueMap()
+                        .putValue(
+                                "integrations",
+                                ValueMap()
+                                        .putValue("foo", ValueMap().putValue("appToken", "foo_token"))
+                                        .putValue(
+                                                "bar",
+                                                ValueMap()
+                                                        .putValue("appToken", "foo_token")
+                                                        .putValue("trackAttributionData", true)))
+        builder =
+                Analytics.Builder(RuntimeEnvironment.application, "write_key")
+                        .defaultProjectSettings(projectSettings)
+                        .use(
+                                object : Integration.Factory {
+                                    override fun create(settings: ValueMap, analytics: Analytics): Integration<*>? {
+                                        return integrationFoo
+                                    }
+
+                                    override fun key(): String {
+                                        return "foo"
+                                    }
+                                })
+                        .use(
+                                object : Integration.Factory {
+                                    override fun create(settings: ValueMap, analytics: Analytics): Integration<*>? {
+                                        return integrationBar
+                                    }
+
+                                    override fun key(): String {
+                                        return "bar"
+                                    }
+                                })
+                        .executor(MoreExecutors.newDirectExecutorService())
+    }
+
+    /** Edge Function Middleware Tests **/
+    @Test
+    @Throws(Exception::class)
+    fun edgeFunctionMiddlewareCanExist() {
+
+        val analytics = builder
+                .useEdgeFunctionMiddleware(Mockito.mock(JSMiddleware::class.java))
+                .build()
+
+        analytics.track("foo")
+
+        val privateEdgeFunctions: Field = Analytics::class.java.getDeclaredField("edgeFunctionMiddleware")
+        Assertions.assertThat(privateEdgeFunctions).isNotNull
+    }
+
+    @Test
+    @Throws(Exception::class)
+    fun edgeFunctionMiddlewareOverwrites() {
+
+        try {
+            val analytics = builder
+                    .useEdgeFunctionMiddleware(Mockito.mock(JSMiddleware::class.java))
+                    .useSourceMiddleware(
```

Comment:
just wanna confirm that we are ok with letting users specify both without raising an exception in this scenario